File: src/weapons/Weapon.js
import { upgradedDef, upgradeCost, UPGRADE_LEVELS } from './WeaponDefs.js';
import { Audio } from '../core/AudioEngine.js';
import { clamp } from '../core/Util.js';

/** Runtime state for one carried weapon. */
export class Weapon {
  constructor(baseDef, upgradeLevel = 0) {
    this.base = baseDef;
    this.upgradeLevel = upgradeLevel;
    this.def = upgradedDef(baseDef, upgradeLevel);

    this.mag = this.def.mag;
    this.reserve = this.def.reserve;
    this.cooldown = 0;
    this.reloading = false;
    this.reloadT = 0;
    this.reloadTotal = 0;
    this.shellsLoaded = 0;
    this.triggerHeld = false;
    this.burstLeft = 0;
    this.burstCooldown = 0;
    this.heat = 0;              // drives recoil climb
    this.shotsThisTrigger = 0;
  }

  get id() { return this.base.id; }
  get name() { return this.def.name; }
  get isEmpty() { return this.mag <= 0; }
  get isFull() { return this.mag >= this.def.mag; }
  get canReload() { return !this.reloading && this.mag < this.def.mag && this.reserve > 0; }
  get fireInterval() { return 60 / this.def.rpm; }
  get maxUpgrade() { return UPGRADE_LEVELS.length; }
  get nextUpgradeCost() { return upgradeCost(this.upgradeLevel); }

  upgrade() {
    if (this.upgradeLevel >= UPGRADE_LEVELS.length) return false;
    this.upgradeLevel++;
    this.def = upgradedDef(this.base, this.upgradeLevel);
    this.mag = this.def.mag;
    this.reserve = this.def.reserve;
    this.reloading = false;
    return true;
  }

  refillAmmo() {
    this.reserve = this.def.reserve;
    this.mag = this.def.mag;
    this.reloading = false;
  }

  addAmmo(fraction = 1) {
    this.reserve = clamp(this.reserve + Math.round(this.def.reserve * fraction), 0, this.def.reserve);
  }

  /** Returns true if the trigger state allows a shot this frame. */
  wantsToFire(held, pressed) {
    if (this.reloading || this.cooldown > 0) return false;
    if (this.def.fireMode === 'auto') return held;
    if (this.def.fireMode === 'burst') return pressed && this.burstCooldown <= 0;
    return pressed;
  }

  /** Consumes a round. Returns false when the magazine is dry. */
  consume() {
    if (this.mag <= 0) return false;
    this.mag--;
    this.cooldown = this.fireInterval;
    this.heat = Math.min(1, this.heat + 0.16);
    this.shotsThisTrigger++;
    return true;
  }

  startReload(speedMult = 1) {
    if (!this.canReload) return false;
    this.reloading = true;
    this.shellsLoaded = 0;
    const empty = this.mag <= 0;
    this.reloadTotal = (empty ? this.def.emptyReloadTime : this.def.reloadTime) / speedMult;
    this.reloadT = this.reloadTotal;
    Audio.play('reload', { stage: 0, volume: 0.9 });
    return true;
  }

  cancelReload() {
    if (!this.reloading) return;
    this.reloading = false;
    this.reloadT = 0;
  }

  update(dt, speedMult = 1) {
    this.cooldown = Math.max(0, this.cooldown - dt);
    this.burstCooldown = Math.max(0, this.burstCooldown - dt);
    this.heat = Math.max(0, this.heat - dt * 1.6);

    if (!this.reloading) return;
    this.reloadT -= dt;

    if (this.def.shellReload) {
      // Shotgun-style: one shell per cycle, interruptible.
      if (this.reloadT <= 0) {
        this.mag++;
        this.reserve--;
        this.shellsLoaded++;
        Audio.play('reload', { stage: 1, volume: 0.8 });
        if (this.mag >= this.def.mag || this.reserve <= 0) {
          this.reloading = false;
        } else {
          this.reloadTotal = this.def.reloadTime / speedMult;
          this.reloadT = this.reloadTotal;
        }
      }
      return;
    }

    // Magazine reload: two audible stages then the round transfer.
    const p = 1 - this.reloadT / this.reloadTotal;
    if (p > 0.45 && !this._stage1) { this._stage1 = true; Audio.play('reload', { stage: 1, volume: 0.85 }); }
    if (p > 0.85 && !this._stage2) { this._stage2 = true; Audio.play('reload', { stage: 2, volume: 0.7 }); }
    if (this.reloadT <= 0) {
      const need = this.def.mag - this.mag;
      const take = Math.min(need, this.reserve);
      this.mag += take;
      this.reserve -= take;
      this.reloading = false;
      this._stage1 = false;
      this._stage2 = false;
    }
  }

  get reloadProgress() {
    return this.reloading ? clamp(1 - this.reloadT / this.reloadTotal, 0, 1) : 1;
  }
}

File: unity/Rotgrid/Assets/Scripts/Weapons/Weapon.cs
using UnityEngine;

namespace Rotgrid
{
    /// <summary>Runtime state for one carried weapon.</summary>
    public class Weapon
    {
        public readonly WeaponDef baseDef;
        public WeaponDef def;
        public int upgradeLevel;

        public int mag, reserve;
        public float cooldown;
        public bool reloading;
        public float reloadT, reloadTotal;
        public float burstCooldown;
        public float heat;
        bool _stage1, _stage2;

        public Weapon(WeaponDef d, int level = 0)
        {
            baseDef = d;
            upgradeLevel = level;
            def = Weapons.Upgraded(d, level);
            mag = def.mag;
            reserve = def.reserve;
        }

        public string Id { get { return baseDef.id; } }
        public string Name { get { return def.name; } }
        public bool IsEmpty { get { return mag <= 0; } }
        public bool CanReload { get { return !reloading && mag < def.mag && reserve > 0; } }
        public float FireInterval { get { return def.FireInterval; } }
        public int NextUpgradeCost { get { return Weapons.UpgradeCost(upgradeLevel); } }
        public float ReloadProgress { get { return reloading ? Mathf.Clamp01(1f - reloadT / Mathf.Max(0.01f, reloadTotal)) : 1f; } }

        public bool Upgrade()
        {
            if (upgradeLevel >= Weapons.Tiers.Length) return false;
            upgradeLevel++;
            def = Weapons.Upgraded(baseDef, upgradeLevel);
            mag = def.mag;
            reserve = def.reserve;
            reloading = false;
            return true;
        }

        public void RefillAmmo()
        {
            reserve = def.reserve;
            mag = def.mag;
            reloading = false;
        }

        /// <summary>True when the trigger state allows a shot this frame.</summary>
        public bool WantsToFire(bool held, bool pressed)
        {
            if (reloading || cooldown > 0f) return false;
            if (def.fireMode == FireMode.Auto) return held;
            if (def.fireMode == FireMode.Burst) return pressed && burstCooldown <= 0f;
            return pressed;
        }

        /// <summary>Consumes a round. Returns false when the magazine is dry.</summary>
        public bool Consume()
        {
            if (mag <= 0) return false;
            mag--;
            cooldown = FireInterval;
            heat = Mathf.Min(1f, heat + 0.16f);
            return true;
        }

        public bool StartReload(float speedMult)
        {
            if (!CanReload) return false;
            reloading = true;
            bool empty = mag <= 0;
            reloadTotal = (empty ? def.emptyReloadTime : def.reloadTime) / speedMult;
            reloadT = reloadTotal;
            _stage1 = _stage2 = false;
            Game.Audio.Play(Sfx.Reload1, 0.9f);
            return true;
        }

        public void CancelReload()
        {
            reloading = false;
            reloadT = 0f;
        }

        public void Update(float dt, float speedMult)
        {
            cooldown = Mathf.Max(0f, cooldown - dt);
            burstCooldown = Mathf.Max(0f, burstCooldown - dt);
            heat = Mathf.Max(0f, heat - dt * 1.6f);

            if (!reloading) return;
            reloadT -= dt;

            if (def.shellReload)
            {
                // Shotgun-style: one shell per cycle, interruptible.
                if (reloadT <= 0f)
                {
                    mag++;
                    reserve--;
                    Game.Audio.Play(Sfx.Reload2, 0.8f);
                    if (mag >= def.mag || reserve <= 0) reloading = false;
                    else
                    {
                        reloadTotal = def.reloadTime / speedMult;
                        reloadT = reloadTotal;
                    }
                }
                return;
            }

            // Magazine reload: two audible stages then the round transfer.
            float p = ReloadProgress;
            if (p > 0.45f && !_stage1) { _stage1 = true; Game.Audio.Play(Sfx.Reload2, 0.85f); }
            if (p > 0.85f && !_stage2) { _stage2 = true; Game.Audio.Play(Sfx.Reload3, 0.7f); }
            if (reloadT <= 0f)
            {
                int need = def.mag - mag;
                int take = Mathf.Min(need, reserve);
                mag += take;
                reserve -= take;
                reloading = false;
                _stage1 = _stage2 = false;
            }
        }
    }
}
